common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java

```diff
@@ -225,76 +236,117 @@ public String encrypt(final String clearText)
-                throw e;
-            }
+        throw new GeneralSecurityException(ErrorStrings.DECRYPTION_FAILED);
+    }
 
-            // We are the broker... try with the other broker's package name
-            final String nextPackageToUse =
-                    packageName.equals(AZURE_AUTHENTICATOR_APP_PACKAGE_NAME)
-                            ? COMPANY_PORTAL_APP_PACKAGE_NAME
-                            : AZURE_AUTHENTICATOR_APP_PACKAGE_NAME;
+    public boolean isEncryptedWithUserDefinedKey(@NonNull final String data) {
+        final String methodName = "isEncryptedWithUserDefinedKey";
+
+        final byte[] bytes;
+        try {
+            bytes = getByteArrayFromEncryptedBlob(data);
+        } catch (IllegalArgumentException e) {
+            Logger.error(TAG + methodName, "This data is not an encrypted blob.", e);
+            return false;
+        }
 
-            Logger.warn(
-                    TAG + methodName,
-                    "We are the broker. Retrying with package name: "
-                            + nextPackageToUse
+        try {
+            final String keyVersion = new String(
+                    bytes,
+                    0,
+                    KEY_VERSION_BLOB_LENGTH,
+                    AuthenticationConstants.ENCODING_UTF8
             );
 
-            return decryptWithPackageName(bytes, nextPackageToUse);
+            return VERSION_USER_DEFINED.equalsIgnoreCase(keyVersion);
+        } catch (UnsupportedEncodingException e) {
+            Logger.error(TAG + methodName, "Failed to extract keyVersion.", e);
         }
+
+        return false;
     }
 
-    private String decryptWithPackageName(final byte[] bytes,
-                                          @NonNull final String packageName)
-            throws GeneralSecurityException, IOException {
-        // get key version used for this data. If user upgraded to different
-        // API level, data needs to be updated
-        final String keyVersion = new String(
-                bytes,
-                0,
-                KEY_VERSION_BLOB_LENGTH,
-                AuthenticationConstants.ENCODING_UTF8
+    private byte[] getByteArrayFromEncryptedBlob(@NonNull final String encryptedBlob) {
+        int encodeVersionLength = encryptedBlob.charAt(0) - 'a';
+        validateEncodeVersion(encryptedBlob, encodeVersionLength);
+
+        return Base64.decode(
+                encryptedBlob.substring(1 + encodeVersionLength),
+                Base64.DEFAULT
         );
+    }
+
+    /**
+     * Get all the key type that could be potential candidates for decryption.
+     **/
+    private List<KeyType> initializeDecryptionKeyTypeList(@NonNull final String encryptedBlob,
+                                                          @NonNull final String packageName) {
+        List<KeyType> keyTypeList = new ArrayList<>();
+
+        if (isEncryptedWithUserDefinedKey(encryptedBlob)) {
```

Comment:
Is it useful to include logging for each of these conditions?

Reply:
This is already logged inside decrypt() for loop, when the decryption fails. I think that should be enough.